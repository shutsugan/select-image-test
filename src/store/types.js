export const FETCH_IMAGES = "FETCH_IMAGES";
export const FETCH_FAILED = "FETCH_FAILED";
export const SELECT_IMAGE = "SELECT_IMAGE";
export const TOGGLE_LOADING = "TOGGLE_LOADING";
