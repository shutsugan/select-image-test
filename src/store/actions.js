import axios from "axios";
import {
  FETCH_IMAGES,
  FETCH_FAILED,
  SELECT_IMAGE,
  TOGGLE_LOADING,
} from "./types";

export const getImages = (images) => ({
  type: FETCH_IMAGES,
  payload: { images },
});

export const selectImage = (image) => ({
  type: SELECT_IMAGE,
  payload: { image },
});

export const toggleLoading = (status) => ({
  type: TOGGLE_LOADING,
  payload: { status },
});

export const setError = () => ({
  type: FETCH_FAILED,
});

export const fetchImages = (url) => async (dispatch) => {
  try {
    dispatch(toggleLoading(true));
    const { data } = await axios.get(url);

    dispatch(getImages(data || []));
    dispatch(toggleLoading(false));
  } catch (err) {
    dispatch(setError());
  }
};
