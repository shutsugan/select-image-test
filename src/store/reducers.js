import {
  FETCH_IMAGES,
  FETCH_FAILED,
  SELECT_IMAGE,
  TOGGLE_LOADING,
} from "./types";

export const initialState = {
  images: [],
  selectedImage: null,
  previousSelectedImage: null,
  loading: false,
  error: null,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_IMAGES: {
      const { images } = action.payload;

      return {
        ...state,
        images,
      };
    }
    case SELECT_IMAGE: {
      const { image } = action.payload;

      return {
        ...state,
        selectedImage: image,
        previousSelectedImage: state.selectedImage,
      };
    }
    case TOGGLE_LOADING: {
      const { status } = action.payload;

      return {
        ...state,
        loading: status,
      };
    }
    case FETCH_FAILED: {
      return {
        ...state,
        error: true,
      };
    }
    default:
      return state;
  }
};

export default reducer;
