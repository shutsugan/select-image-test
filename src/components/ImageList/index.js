import React from "react";
import { useDispatch } from "react-redux";

import { selectImage } from "../../store/actions";

const ImageList = ({ images = [] }) => {
  const dispatch = useDispatch();

  const handleSelectImage = (image) => dispatch(selectImage(image));

  return (
    <div className="images-list grid">
      {images.map((image) => (
        <div
          key={image}
          className="image-item flex shadow"
          onClick={() => handleSelectImage(image)}
        >
          <img className="image" src={image} alt="" />
        </div>
      ))}
    </div>
  );
};

export default ImageList;
