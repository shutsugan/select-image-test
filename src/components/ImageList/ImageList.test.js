import React from "react";
import { render } from "@testing-library/react";

import { Provider } from "react-redux";
import store from "../../store";

import ImageList from "./";

test("generate ImageList component snapshot", () => {
  const { container } = render(
    <Provider store={store}>
      <ImageList image={[]} />
    </Provider>
  );

  expect(container).toMatchSnapshot();
});
