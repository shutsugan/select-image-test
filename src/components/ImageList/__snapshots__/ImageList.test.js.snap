// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate ImageList component snapshot 1`] = `
<div>
  <div
    class="images-list grid"
  />
</div>
`;
